langchain/utilities/arxiv_search.py

```diff
@@ -0,0 +1,138 @@
+            "submittedDate": SortCriterion.SubmittedDate,
+        }
+        sort_order_mapping = {
+            "ascending": SortOrder.Ascending,
+            "descending": SortOrder.Descending,
+        }
+
+        values["sort_by"] = sort_by_mapping[values["sort_by"]]
+        values["sort_order"] = sort_order_mapping[values["sort_order"]]
+
+        try:
+            import arxiv
+        except ImportError:
+            raise ImportError(
+                "arxiv is not installed. " "Please install it with `pip install arxiv`"
+            )
+        return values
+
+    def run(self, query: str) -> str:
+        """Run query through ArXiv Search and parse result."""
+        papers = self.results(query)
+
+        # Representation of a list of papers in a string:
+        # """
+        # Title: a-title
+        # Authors: a, b, c
+        # Abstract: an-abstract
+        #
+        #
+        # Title: another-title...
+        # """
+        if len(papers) == 0:
+            return "No good Arxiv Search Result was found"
+
+        results = ""
+        for paper in papers:
+            result = f"Title: {paper['title'].strip()}\n"
+            result += f"Authors: {_get_authors_str(paper['authors']).strip()}\n"
+            result += f"Abstract: {paper['abstract'].strip()}\n"
+            result += "\n\n"
+            results += result.strip()
+
+        return results
+
+    def results(self, query: str) -> List[dict]:
+        """Return results from ArXiv."""
+        return self._parse_results(self._arxiv_search_results(query))
+
+    def _parse_results(self, results: Iterator) -> List[dict]:
+        papers = []
+
+        try:
+            result = next(results)
+
+            paper_id = result.get_short_id()
+            paper_title = result.title
+            paper_url = result.entry_id
+            paper_abstract = result.summary.replace("\n", " ")
+            paper_authors = [str(author) for author in result.authors]
+            paper_first_author = _get_authors_str(result.authors, first_author=True)
+            primary_category = result.primary_category
+            publish_time = result.published.date()
+            update_time = result.updated.date()
+            comments = result.comment
+
+            paper = {
+                "id": paper_id,
+                "title": paper_title,
+                "url": paper_url,
+                "abstract": paper_abstract,
+                "authors": paper_authors,
+                "first_author": paper_first_author,
+                "primary_category": primary_category,
+                "publish_time": publish_time,
+                "update_time": update_time,
+                "comments": comments,
+            }
+            papers.append(paper)
+        except (
+            StopIteration
+        ):  # See: https://docs.python.org/3/library/stdtypes.html#iterator.__next__
+            pass  # No results found
+
+        return papers
```

Comment:
I think you can eliminate much of this code by using the `Search(...).results()` generator directly, rather than wrapping it in a generator (`_parse_results`) converting the results from objects to maps.

If you *do* want to do that conversion, you might consider using the `__dict__ ` built-in: `result. __dict__ `.

Note: this rewrite of `run` assumes the suggestion I made below in this file (`_arxiv_search_results` → `_search`).

```suggestion
def run(self, query: str) -> str:
        """Run query through ArXiv Search and parse result."""
        # papers = self.results(query)

        papers = []
        for paper in self._search(query).results():
            papers.append("\n".join([
                f'Title: {paper.title.strip()}',
                f'Authors: {", ".join([a.name for a in paper.authors])}',
                f'Abstract: {paper.summary.strip()}'
            ]))

        if not papers:
            return "No good arXiv Search result was found"
        
        return "\n\n".join(papers)
```